### PR TITLE
Pv2 2809 net negative payments metrics

### PR DESCRIPTION
**Pull request by AkshayBanifatsyevichChugh**

### CODE REVIEW

File: src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs [hunk 1]
```diff
@@ -183,10 +205,41 @@ public void AddInLearningCount(ProviderInLearningTotal inLearningTotal)
...
+            negativeEarnings = new NegativeEarningsContractTypeAmounts();
+
+            if (providerLearnerNegativeDcEarnings == null || providerLearnerNegativeDcEarnings.Count == 0) return;
+
+            var distinctLearnerUlns = providerLearnerNegativeDcEarnings.Select(x => x.Uln).Distinct().ToList();
+
+            foreach (var uln in distinctLearnerUlns)
+            {
+                //if learner has payments, then do not add negative earnings to provider summary
+                var learnerPayments = providerLearnerPayments?.Where(x => x.LearnerUln == uln && x.Total > 0m).ToList();
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs:219]: check if  Total != 0m

File: src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs [hunk 2]
```diff
@@ -183,10 +205,41 @@ public void AddInLearningCount(ProviderInLearningTotal inLearningTotal)
...
+                if (learnerPayments != null && learnerPayments.Count != 0) continue;
+
+                //if learner has data locks, then do not add negative earnings to provider summary
+                var learnerDataLocks = providerLearnerDataLockedEarnings?.Where(x => x.LearnerUln == uln).ToList();
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs:223]: check if  Total != 0m

File: src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs [hunk 3]
```diff
@@ -183,10 +205,41 @@ public void AddInLearningCount(ProviderInLearningTotal inLearningTotal)
...
+                            negativeEarnings.ContractType1 = negativeEarnings.ContractType1.GetValueOrDefault() + x.NegativeEarningsTotal;
+                            break;
+
+                        case ContractType.Act2:
+                            negativeEarnings.ContractType2 = negativeEarnings.ContractType2.GetValueOrDefault() + x.NegativeEarningsTotal;
+                            break;
+                    }
+                });
+            }
         }
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs:242]: This method can be simplified as bellow if the duplicates is not an issue

            negativeEarnings = new NegativeEarningsContractTypeAmounts();

            if (providerLearnerNegativeDcEarnings == null || providerLearnerNegativeDcEarnings.Count == 0) return;

            foreach (var learner in providerLearnerNegativeDcEarnings)
            {
                //if learner has payments, then do not add negative earnings to provider summary
                var learnerPayments = providerLearnerPayments?.Where(x => x.LearnerUln == learner.Uln && x.Total > 0m).ToList();
                if (learnerPayments != null && learnerPayments.Count != 0) continue;

                //if learner has data locks, then do not add negative earnings to provider summary
                var learnerDataLocks = providerLearnerDataLockedEarnings?.Where(x => x.LearnerUln == learner.Uln).ToList();
                if (learnerDataLocks != null && learnerDataLocks.Count != 0) continue;

                switch (learner.ContractType)
                {
                    case ContractType.Act1:
                        negativeEarnings.ContractType1 =+ learner.NegativeEarningsTotal;
                        break;

                    case ContractType.Act2:
                        negativeEarnings.ContractType2 =+ learner.NegativeEarningsTotal;
                        break;
                }
            }

File: src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs [hunk 4]
```diff
@@ -183,10 +205,41 @@ public void AddInLearningCount(ProviderInLearningTotal inLearningTotal)
...
-        public void AddNegativeEarnings(List<ProviderNegativeEarningsTotal> providerNegativeEarningsTotal)
+        private void CalculateNegativeEarnings()
         {
-            negativeEarnings.ContractType1 = providerNegativeEarningsTotal.FirstOrDefault(x => x.ContractType == ContractType.Act1)?.NegativeEarningsTotal ?? 0m;
-            negativeEarnings.ContractType2 = providerNegativeEarningsTotal.FirstOrDefault(x => x.ContractType == ContractType.Act2)?.NegativeEarningsTotal ?? 0m;
+            negativeEarnings = new NegativeEarningsContractTypeAmounts();
+
+            if (providerLearnerNegativeDcEarnings == null || providerLearnerNegativeDcEarnings.Count == 0) return;
+
+            var distinctLearnerUlns = providerLearnerNegativeDcEarnings.Select(x => x.Uln).Distinct().ToList();
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs:214]: are you sure there are duplicates? if there ware duplicate's why there are duplicates?

**Review by hardik-desai-kcl — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.Payments.Monitoring.Metrics.Application/Infrastructure/Ioc/MetricsModule.cs [hunk 5]
```diff
@@ -1,6 +1,7 @@
 ﻿using Autofac;
 using Microsoft.EntityFrameworkCore;
 using SFA.DAS.Payments.Core.Configuration;
+using SFA.DAS.Payments.Monitoring.Metrics.Application.PeriodEnd;
 using SFA.DAS.Payments.Monitoring.Metrics.Application.Submission;
```
[the remark this answers is not in the document]
Comment (AkshayBanifatsyevichChugh) [reply src/SFA.DAS.Payments.Monitoring.Metrics.Application/Infrastructure/Ioc/MetricsModule.cs]: removed

File: src/SFA.DAS.Payments.Monitoring.Metrics.Model/ProviderLearnerDataLockEarningsTotal.cs [hunk 6]
```diff
@@ -1,8 +1,9 @@
 ﻿namespace SFA.DAS.Payments.Monitoring.Metrics.Model
 {
-    public class ProviderTotal
+    public class ProviderLearnerDataLockEarningsTotal
```
[the remark this answers is not in the document]
Comment (AkshayBanifatsyevichChugh) [reply src/SFA.DAS.Payments.Monitoring.Metrics.Model/ProviderLearnerDataLockEarningsTotal.cs]: change applied

File: src/SFA.DAS.Payments.Monitoring.Metrics.Model/ProviderLearnerContractTypeAmounts.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿namespace SFA.DAS.Payments.Monitoring.Metrics.Model
+{
+    public class ProviderLearnerContractTypeAmounts : ProviderContractTypeAmounts
```
[the remark this answers is not in the document]
Comment (AkshayBanifatsyevichChugh) [reply src/SFA.DAS.Payments.Monitoring.Metrics.Model/ProviderLearnerContractTypeAmounts.cs]: change applied

File: src/SFA.DAS.Payments.Monitoring.Metrics.Model/ProviderLearnerNegativeEarningsTotal.cs [hunk 8]
```diff
@@ -1,14 +1,12 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Text;
-using SFA.DAS.Payments.Model.Core.Entities;
+﻿using SFA.DAS.Payments.Model.Core.Entities;
 
 namespace SFA.DAS.Payments.Monitoring.Metrics.Model
 {
-    public class ProviderNegativeEarningsTotal
+    public class ProviderLearnerNegativeEarningsTotal
```
[the remark this answers is not in the document]
Comment (AkshayBanifatsyevichChugh) [reply src/SFA.DAS.Payments.Monitoring.Metrics.Model/ProviderLearnerNegativeEarningsTotal.cs]: change applied

File: src/SFA.DAS.Payments.Monitoring.Metrics.Application/PeriodEnd/PeriodEndMetricsRepository.cs [hunk 9]
```diff
@@ -9,14 +9,17 @@
...
 using System.Threading;
 using System.Threading.Tasks;
+using SFA.DAS.Payments.Core;
 
 namespace SFA.DAS.Payments.Monitoring.Metrics.Application.PeriodEnd
 {
     public interface IPeriodEndMetricsRepository
     {
         Task<List<ProviderTransactionTypeAmounts>> GetTransactionTypesByContractType(short academicYear, byte collectionPeriod, CancellationToken cancellationToken);
+        Task<List<ProviderLearnerContractTypeAmounts>> GetPaymentAmountsForLearnersByContractType(List<long> learnerUlns, short academicYear, CancellationToken cancellationToken);
```
[the remark this answers is not in the document]
Comment (AkshayBanifatsyevichChugh) [reply src/SFA.DAS.Payments.Monitoring.Metrics.Application/PeriodEnd/PeriodEndMetricsRepository.cs]: change applied

File: src/SFA.DAS.Payments.Monitoring.Metrics.Application/PeriodEnd/PeriodEndMetricsRepository.cs [hunk 10]
```diff
@@ -9,14 +9,17 @@
...
         Task<List<ProviderFundingSourceAmounts>> GetFundingSourceAmountsByContractType(short academicYear, byte collectionPeriod, CancellationToken cancellationToken);
         Task<List<ProviderFundingLineTypeAmounts>> GetDataLockedEarningsTotals(short academicYear, byte collectionPeriod, CancellationToken cancellationToken);
+        Task<List<ProviderLearnerDataLockEarningsTotal>> GetDataLockedAmountsForLearners(List<long> learnerUlns, short academicYear, byte collectionPeriod, CancellationToken cancellationToken);
```
[the remark this answers is not in the document]
Comment (AkshayBanifatsyevichChugh) [reply src/SFA.DAS.Payments.Monitoring.Metrics.Application/PeriodEnd/PeriodEndMetricsRepository.cs]: change applied

File: src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs (hunk 1 above)
Comment (AkshayBanifatsyevichChugh) [reply src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs]: change applied

File: src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs (hunk 2 above)
Comment (AkshayBanifatsyevichChugh) [reply src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs]: change applied

File: src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs (hunk 3 above)
Comment (AkshayBanifatsyevichChugh) [reply src/SFA.DAS.Payments.Monitoring.Metrics.Domain/PeriodEnd/PeriodEndProviderSummary.cs:242]: change applied

**Review by hardik-desai-kcl — DISMISSED** [review]

**Review by hardik-desai-kcl — DISMISSED** [review]

**Review by hardik-desai-kcl — APPROVED** [review]

**Pull request merged**